### ISSUE TITLE
Incorrect backward for smooth_l1_loss when beta = 0 on GPU

### ISSUE THREAD

**Issue by CedricPicron** [pytorch/pytorch#96813]
### 🐛 Describe the bug

**Bug description:**

Incorrect backward (i.e. NaN gradients) for smooth_l1_loss on GPU when `beta = 0` and when inputs and targets are nearly identical (see different `eps` values in code below). The gradient is correct when computed on CPU. I didn't have this issue when using PyTorch1.10 (and I believe it also worked correctly for PyTorch1.12, but I'm not 100% certain).

**Example to reproduce:**
```python
import torch
import torch.nn.functional as F

for eps in (0.000, 1e-8, 1e-7, 1e-6, 1e-5):
    input = torch.randn(300, 256, requires_grad=True, device='cuda')
    target = input.detach() + eps

    loss = F.smooth_l1_loss(input, target, beta=0.0, reduction='sum')
    loss.backward()

    if eps == 0:
        print(f"{eps:.3f}: {input.grad.mean().item()}")
    else:
        print(f"{eps}: {input.grad.mean().item()}")
```
**Output:**
```
0.000: nan
1e-08: nan
1e-07: nan
1e-06: -1.0
1e-05: -1.0
```
**Expected output:**
```
0.000: -1.0
1e-08: -1.0
1e-07: -1.0
1e-06: -1.0
1e-05: -1.0
```

### Versions

PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Fedora Linux 37 (Thirty Seven) (x86_64)
GCC version: (cpicron-ruchba) 9.3.0
Clang version: 15.0.7 (Fedora 15.0.7-1.fc37)
CMake version: version 3.26.0-rc5

Python version: 3.10 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050 Ti
Nvidia driver version: 525.89.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torch-cluster==1.6.0
[pip3] torch-geometric==2.2.0
[pip3] torch-scatter==2.1.0
[pip3] torch-sparse==0.6.16
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pyg                       2.2.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch                   1.13.1          py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cluster           1.6.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch-cuda              11.7                 h67b0de4_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-scatter           2.1.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch-sparse            0.6.16          py310_torch_1.13.0_cu117    pyg
[conda] torchvision               0.14.1              py310_cu117    pytorch

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki @ngimel

**Comment by soulitzer** [comment]
You should expect to get some weird results since you're using float32. Some of the NaNs go away if you use float64 or use zeros instead of randn. 

Maybe we should get rid of the NaNs though, i.e. do what l1_loss is doing. Maybe we could just test for beta=0 and call l1_loss?

Note that the expected output for eps 0.0 per https://pytorch.org/docs/stable/notes/autograd.html#gradients-for-non-differentiable-functions should be 0. Since l1_loss is convex and non-differentiable at that point, we should choose the min-norm sub-gradient

**Comment by CedricPicron** [comment]
I think that NaNs should indeed be avoided at all cost, since this corrupts the whole training process (everything including model parameters become NaN) and given it is hard to debug (took me a couple of hours) as this issue might only occur once every 25k iterations.

This behavior is a regression compared to previous PyTorch versions such as PyTorch 1.10, which gives following output:

```
0.000: 0.0
1e-08: -0.1987239569425583
1e-07: -0.9543099403381348
1e-06: -1.0
1e-05: -1.0
```

Would it be possible to go back to this behavior (or any other behavior avoiding the NaNs)? I quickly tried to find out what changed compared to previous versions, but didn't find anything conclusive.

**Comment by soulitzer** [comment]
Marking as high priority because of regression. (Code doesn't produce NaNs on 1.12.1 either.)

**Comment by lezcano** [comment]
Couldn't this be masked manually by the user if they want to recover the behaviour?

**Comment by soulitzer** [comment]
Yeah, `input.register_hook(lambda x: torch.nan_to_num(x, nan=0.0))` could help recover the behavior here. Just using l1_loss works too.

**Comment by ngimel** [comment]
However, I would second @soulitzer's point of not expecting correct results when subtracting two very close numbers in fp32, slightly modifying the script above to do `-eps` instead of `+eps` yields questionable results even on cpu (because of the order in which masking is done):
```
In [6]: for eps in (0.000, 1e-8, 1e-7, 1e-6, 1e-5):
   ...:     input = torch.randn(300, 256, requires_grad=True, device='cpu')
   ...:     target = input.detach() - eps
   ...: 
   ...:     loss = F.smooth_l1_loss(input, target, beta=0.0, reduction='sum')
   ...:     loss.backward()
   ...: 
   ...:     if eps == 0:
   ...:         print(f"{eps:.3f}: {input.grad.mean().item()}")
   ...:     else:
   ...:         print(f"{eps}: {input.grad.mean().item()}")
   ...: 
0.000: -1.0
1e-08: -0.6043750047683716
1e-07: 0.9104427099227905
1e-06: 1.0
1e-05: 1.0
```

**Comment by ngimel** [comment]
According to our self-imposed rules, non-differentiable functions should produce lowest norm subgradients (which would be 0 in this case), so we are in violation.

**Comment by CedricPicron** [comment]
Mostly for _historical_ reasons. First, I used it with some `beta > 0`, but then switched to `beta = 0` as this setting worked equally well. Didn't change since. 

The second minor reason is that I use my own SmoothL1Loss module which is the nn.SmoothL1Loss that additionally allows a loss weight, i.e. `loss = weight * F.smooth_l1_loss(input, target, beta, reduction)`. By doing this for the SmoothL1Loss, I only need to do this once, and not for the SmoothL1Loss and L1Loss separately.

**Comment by ngimel** [comment]
So for a quick workaround you can dispatch to `l1_loss` inside your module if `beta` is 0, but yeah, I agree we should fix at least nan issue, even if small `input-target` would still have inaccurate results.

**Comment by albanD** [comment]
From a quick chat, the way to go forward here, we should update the core implementation https://github.com/pytorch/pytorch/blob/3f4090652cd827ebb449789f6e49aff6b08dbe33/torch/nn/functional.py#L3208 to call into `l1_loss` when `beta` is 0. Would you be able to send a PR for that and a one-off test in test_nn to ensure we provide the right result for these edge cases.

**Comment by lezcano** [comment]
I remember trying to do that at some point and hitting some pesky errors or smth when I updated the l1_loss implementation. I don't remember what happened exactly, but I remember having this same idea and then not doing that dispatch. But perhaps things are fine now?

**Comment by albanD** [comment]
Ho that is interesting. I only suggested to do this at the python level btw ;)

**Comment by CedricPicron** [comment]
Yes, I can do it. I'll work on it tomorrow.
